### ISSUE TITLE
Research vets-api retry / queuing capability

### ISSUE THREAD

**Issue by batemapf** [department-of-veterans-affairs/va.gov-team#41359]
CARMA team wants to implement queue/retries at the MuleSoft API layer (middleware b/w vets-api and the CARMA/Salesforce) to handle cases where CARMA/SF cannot accept a submission because _{reasons}_. Rather than investing in that implementation and cost, how feasible is it to just add retries at vets-api + some kind of expiration action (e.g. after x retries, generate PDF, send to central mail). Given past experience with reconfiguring vets-api to work with MuleSoft (apparently was supposed to be easy, but end up sucking up tons of time our end), maybe this is faster/easier in the long run.

**Comment by lihanli** [comment]
# Pros and cons of Mulesoft

### Pros
We don't have to build and maintain the feature ourselves

### Cons
Reduced visibility into errors that happen
Increased latency
Reliant on another team to fix issues that occur
Using a middleware API introduces a new system that can cause bugs and security problems

### Previous experience with middleware APIs

I previously worked on another team that used the EVSS claims API, which was based on the same pattern of being a middleware API for the BGS API. The main benefit of the EVSS claims API was that we could pass and receive JSON data to them and they would handle communicating with BGS's SOAP API (an outdated and hard to use API format).

We experienced two major problems with this API. First, there was a huge amount of latency (120s+) with certain calls. EVSS always blamed the latency on BGS but we didn't know for sure who was the actual cause or if it was a combination of both systems.

Second, there was a major security incident that happened with this API. We were using header authentication to tell EVSS what user's claims to retrieve, and then they would send back a cookie for that user (this should not have happened because cookie auth should only be used for browsers not backend to backend APIs). Then the library we were using for http requests would save that cookie and send it back again on the next request (which could have been for a different user). We were always sending the correct header authentication, but the EVSS backend system would prioritize the cookie authentication and ignore the header authentication, therefore some users would see partial details about other user's claims. We were at fault for using a library that would save and resend cookies and they were at fault for having an undisclosed cookie authentication system, based on their API spec they should have only been using header authentication.

TLDR; a security incident occurred that was caused by a combination of EVSS and vets-api problems, it had nothing to do with BGS and could have been avoided by 
not having EVSS as the middleman.

### Recommendation

My recommendation is to implement the queuing on the vets-api side. We already have queuing done for the 1010ez, we can copy the pattern that was used there. Also, PDF generation is already implemented for the 1010cg so it should be easy to send that PDF to the central mail API.

**Comment by batemapf** [comment]
thanks @lihanli, this is excellent. 

i'd like to connect with the CARMA folks and make an offer to handle retries per your recommendation. do you think we should go to them with a description of what kind of retry capability already exists in vets-api, or ask them for a set of requirements (e.g., do retries in this pattern for this long, then do this) and maybe do some customization on our end?

**Comment by lihanli** [comment]
@batemapf I'm not sure who's in charge of the business logic requirements for this feature. If they already figured out the requirements we could take them and implement them, otherwise we would have to figure out the requirements ourselves, maybe by copying the way the 1010ez is doing it and making adjustments as necessary.
